### PR TITLE
Update lookupSimilarEntry to filter superset entries

### LINKED ISSUE #4057: Inconsistent behavior when creating registration entries
Running the following command (with two selectors) for the first time, it creates an entry successfully:
```
./bin/spire-server entry create -selector color:red -selector size:medium  -parentID spiffe://example.org/agent -spiffeID spiffe://example.org/test
Entry ID         : 70e5514c-09e6-4172-9928-e1ac01a18fae
SPIFFE ID        : spiffe://example.org/test
Parent ID        : spiffe://example.org/agent
Revision         : 0
X509-SVID TTL    : default
JWT-SVID TTL     : default
Selector         : color:red
Selector         : size:medium
```
After that, running the following command (with only one of the two selectors above), it returns an error,, even though the selectors are a subset of the selectors in the previous entry:
```
./bin/spire-server entry create -selector color:red  -parentID spiffe://example.org/agent -spiffeID spiffe://example.org/test
Failed to create the following entry (code: AlreadyExists, msg: "similar entry already exists"):
Entry ID         : (none)
SPIFFE ID        : spiffe://example.org/test
Parent ID        : spiffe://example.org/agent
Revision         : 0
X509-SVID TTL    : default
JWT-SVID TTL     : default
Selector         : color:red

Error: failed to create one or more entries
```
However, if running the following (with 3 selectors, 2 of them identical to the first command), it creates the entry successfully.
```
./bin/spire-server entry create -selector color:red -selector size:medium -selector shape:round  -parentID spiffe://example.org/agent -spiffeID spiffe://example.org/test
Entry ID         : 58834c99-ad87-4d05-a742-2ab343b45f1a
SPIFFE ID        : spiffe://example.org/test
Parent ID        : spiffe://example.org/agent
Revision         : 0
X509-SVID TTL    : default
JWT-SVID TTL     : default
Selector         : color:red
Selector         : shape:round
Selector         : size:medium
```
The entry could be created. This is an inconsistent behavior.

### PR DESCRIPTION
**Pull request by MarcosDY**
When trying to create an entry, we are filtering entries that are similar, 
but that filtering is not working as expected and returning entries that has a superset of selectors.
This code prevents that to happens.

**Which issue this PR fixes**
fixes #4057

### CODE REVIEW

File: pkg/server/datastore/sqlstore/sqlstore_test.go [hunk 1]
```diff
@@ -1585,11 +1585,128 @@ func (s *PluginSuite) TestCreateRegistrationEntry() {
...
+			name: "subset selectors",
+			updateEntry: func(e *common.RegistrationEntry) *common.RegistrationEntry {
+				e.Selectors = []*common.Selector{
+					{Type: "a", Value: "1"},
+				}
+				return e
+			},
+		},
+		{
+			name: "with partial selectors",
```
Comment (amoore877) [inline pkg/server/datastore/sqlstore/sqlstore_test.go:1656]:
```suggestion
			name: "with superset selectors",
```
nit on naming

also for an actual test of `partial`, while we're doing sanity testing of this check, could do an intersection, `a:1` and `d:4` and a mutually exclusive `d:4` and `e:5`

File: pkg/server/datastore/sqlstore/sqlstore_test.go [hunk 2]
```diff
@@ -1585,11 +1585,128 @@ func (s *PluginSuite) TestCreateRegistrationEntry() {
...
+		s.assertEntryEqual(s.T(), validRegistrationEntry, registrationEntry, now)
+	}
+}
+
+func (s *PluginSuite) TestCreateOrReturnRegistrationEntry() {
+	now := time.Now().Unix()
+
+	for _, tt := range []struct {
+		name          string
+		updateEntry   func(*common.RegistrationEntry) *common.RegistrationEntry
```
Comment (amoore877) [inline pkg/server/datastore/sqlstore/sqlstore_test.go:1597]:
```suggestion
		cloneAndModifyEntry   func(*common.RegistrationEntry) *common.RegistrationEntry
```
or similar name?
`updateEntry`, especially given Server and DB API names, has different connotations

File: pkg/server/datastore/sqlstore/sqlstore.go [hunk 3]
```diff
@@ -366,7 +366,7 @@ func (ds *Plugin) createOrReturnRegistrationEntry(ctx context.Context,
 	}
 
 	if err = ds.withWriteTx(ctx, func(tx *gorm.DB) (err error) {
-		registrationEntry, err = lookupSimilarEntry(ctx, ds.db, tx, entry)
+		registrationEntry, err = lookupSimilarEntry(ctx, ds.db, ds.log, entry)
```
Comment (azdagron) [inline pkg/server/datastore/sqlstore/sqlstore.go:369]: I think we should be executing this query on the transaction.....

File: pkg/server/datastore/sqlstore/sqlstore_test.go (hunk 2 above)
Comment (MarcosDY) [reply pkg/server/datastore/sqlstore/sqlstore_test.go:1597]: I'll update to modifyEntry, this function is not cloning entry

**Review by amoore877 — APPROVED** [review]

File: pkg/server/datastore/sqlstore/sqlstore_test.go (hunk 1 above)
Comment (amoore877) [reply pkg/server/datastore/sqlstore/sqlstore_test.go]: > also for an actual test of partial, while we're doing sanity testing of this check, could do an intersection, a:1 and d:4 and a mutually exclusive d:4 and e:5

would be nice to do this one still; we'll be a bit sad in the future if we break this functionality in an interesting way that's not caught by unit testing the possible Set permutations

**Review by azdagron — APPROVED** [review]

**Pull request merged**